### PR TITLE
Added an Image logger callback for wandb

### PR DESCRIPTION
**Pull request by sirtris**
Hi,
As mentioned some time ago in issue #127 here is a custom callback that logs a batch of images together with their label and the corresponding prediction to wandb. 

I think it would also be nice to have a callback for object detection and semantic segmentation.
Maybe I could do this in the coming time.

### CODE REVIEW

**Comment by ashleve** [comment]
@sirtris Hi, thanks a lot for the contribution! Let me propose a couple of changes to improve it.

As for the other callbacks, feel free make a PR with your propositions, but the problem is you can't really use them with MNIST example. I don't want to pollute the template with any extra models or datamodules since it would require users to delete more files when starting a new project. 
We could still just add them to `wandb_callbacks.py` though, no problem with that.

File: src/callbacks/wandb_callbacks.py [hunk 1]
```diff
@@ -196,3 +196,40 @@ def on_validation_epoch_end(self, trainer, pl_module):
...
+
+    def on_sanity_check_start(self, trainer, pl_module):
+        self.ready = False
+
+    def on_sanity_check_end(self, trainer, pl_module):
+        """Start executing this callback only after all validation sanity checks end."""
+        self.ready = True
+
+    def on_validation_epoch_end(self, trainer, pl_module):
+        # get a validation batch from the calidation dat loader
```
Comment (ashleve) [inline src/callbacks/wandb_callbacks.py:219]: You need to add here:
```
if self.ready:
```
as it is done in other callbacks. Lightning executes a couple of short sanity checks for validation dataloader at the beginning of the training, which will log extra images if you don't add this line, it also could result in crashing the run if you set more num_samples than available during sanity check.

File: src/callbacks/wandb_callbacks.py [hunk 2]
```diff
@@ -196,3 +196,40 @@ def on_validation_epoch_end(self, trainer, pl_module):
...
+        val_samples = next(iter(trainer.datamodule.val_dataloader()))
+        val_imgs, val_labels = val_samples[0], val_samples[1]
+
+        # run the batch through the network
+        val_imgs = val_imgs.to(device=pl_module.device)
+        logits = pl_module(val_imgs)
+        preds = torch.argmax(logits, axis=-1)
+
+        # log the images as wandb Image
+        trainer.logger.experiment[0].log({
```
Comment (ashleve) [inline src/callbacks/wandb_callbacks.py:229]: Please use `get_wandb_logger()` method from top of the file when accessing logger. This way the callback will also work if you use many different loggers at once, and a proper exception will be thrown when wandb logger is not available.

File: src/callbacks/wandb_callbacks.py (hunk 1 above)
Comment (ashleve) [inline src/callbacks/wandb_callbacks.py:219]:
```suggestion
        # get a validation batch from the validation dataloader
```

File: src/callbacks/wandb_callbacks.py (hunk 2 above)
Comment (ashleve) [reply src/callbacks/wandb_callbacks.py:229]: https://github.com/ashleve/lightning-hydra-template/blob/74eb7c7e57f16607d6b8ae62a627b343248e41a7/src/callbacks/wandb_callbacks.py#L15-L26

File: src/callbacks/wandb_callbacks.py [hunk 3]
```diff
@@ -196,3 +196,40 @@ def on_validation_epoch_end(self, trainer, pl_module):
...
+            'Images':[wandb.Image(x, caption=f'Pred:{pred}, Label:{y}')
```
Comment (ashleve) [inline src/callbacks/wandb_callbacks.py:230]: You could use:
```suggestion
            f"Images/{trainer.logger.experiment[0].name}":[wandb.Image(x, caption=f"Pred:{pred}, Label:{y}")
```
This way we're using `/` character wchich will create separate section in wandb dashboard called "Images"

Comment (sirtris) [reply src/callbacks/wandb_callbacks.py:230]: Seems like good idea, however when I  test it the captions (Pred and Label) are no longer added to the images. 
Do you know why?

Comment (ashleve) [reply src/callbacks/wandb_callbacks.py]: The captions are visible for me when I enter into run view 
image
The previous version looks the same for me in run view:
image
The difference is only in multi experiment dashboard view:
image
But this view always shows only one image from each experiment so it's not very useful either way...
I hope wandb will fix this some day.

File: src/callbacks/wandb_callbacks.py [hunk 4]
```diff
@@ -196,3 +196,44 @@ def on_validation_epoch_end(self, trainer, pl_module):
...
+class ImagePredictionLogger(Callback):
+    """Logs a validation batch and their predictions to wandb.
+       Example adapted from:
+       https://wandb.ai/wandb/wandb-lightning/reports/Image-Classification-using-PyTorch-Lightning--VmlldzoyODk1NzY
+    """
+
+    def __init__(self, num_samples):
+        super().__init__()
+        self.num_samples = num_samples
+
```
Comment (ashleve) [inline src/callbacks/wandb_callbacks.py:210]: This declaration is also needed in init to ensure it will work correctly even if number of validation sanity checks is set to 0 in trainer
```
self.ready = True
```

File: src/callbacks/wandb_callbacks.py [hunk 5]
```diff
@@ -196,3 +196,45 @@ def on_validation_epoch_end(self, trainer, pl_module):
...
             self.targets.clear()
+            
+
+class ImagePredictionLogger(Callback):
+    """Logs a validation batch and their predictions to wandb.
+       Example adapted from:
+       https://wandb.ai/wandb/wandb-lightning/reports/Image-Classification-using-PyTorch-Lightning--VmlldzoyODk1NzY
+    """
+
+    def __init__(self, num_samples):
```
Comment (ashleve) [inline src/callbacks/wandb_callbacks.py:207]:
```suggestion
    def __init__(self, num_samples: int = 8):
```

File: src/callbacks/wandb_callbacks.py (hunk 3 above)
Comment (sirtris) [reply src/callbacks/wandb_callbacks.py]: Ok I checked again on my run view and the captions appear and disappear with the size of the box where the images are placed in.
Should we add a comment somewhere that the box needs to be dragged bigger to see the captions?

Comment (ashleve) [reply src/callbacks/wandb_callbacks.py]: Nah, I would just let it be. Those image views in wandb behave weirdly and I assume they will just fix it over time.

**Review by ashleve — APPROVED** [review]

**Comment by ashleve** [comment]
Okay I will merge that, thanks for the contribution!

**Comment by sirtris** [comment]
Thanks for the improvement tips btw.
Also thanks for your patience. I know some of my commits are a bit messy as this was ma first PR

**Comment by ashleve** [comment]
> Thanks for the improvement tips btw.
> Also thanks for your patience. I know some of my commits are a bit messy as this was ma first PR

No problem, I'm happy I got the chance to practice reviewing!
